lib/logstash/filters/parser.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require 'digest'
+
+# Define a Parser class for processing tokens and generating templates.
+class Parser
+  def initialize(tokens_list, single_dict, double_dict, tri_dict, threshold)
+    @tokens_list = tokens_list
+    @single_dict = single_dict
+    @double_dict = double_dict
+    @tri_dict = tri_dict
+    @threshold = threshold
+
+    @entropy_dict = {}
+  end
+
+  # Helper method to update entropy dictionary
+  def update_entropy_dict(f_score)
+    f_str = f_score.to_s
+    if @entropy_dict.include?(f_str)
+      @entropy_dict[f_str] += 1
+    else
+      @entropy_dict[f_str] = 1
+    end
+  end
+
+  def is_dynamic(tokens, dynamic_index, index)
+    f = 0
+
+    if index.zero?
+      f = 1
+    elsif index == 1
+      singlegram = tokens[index - 1]
+      doublegram = "#{tokens[index - 1]}^#{tokens[index]}"
+
+      f = if @double_dict.include?(doublegram) && @single_dict.include?(singlegram)
+            @double_dict[doublegram].to_f / @single_dict[singlegram]
+          else
+            0
+          end
+
+      update_entropy_dict(f)
+    else
+      if dynamic_index.include?(index - 2)
```

Comment:
I am really confused what `dyanmic_index` is, ping me when you have this commented